### PR TITLE
Changing the minimum supported numpy version to 1.16

### PR DESCRIPTION
**Pull request by bsipocz**
This PR will remove support for numpy <1.16. First commit to do the change, follow-ups to come to do the actual cleanup.

### CODE REVIEW

**Comment by bsipocz** [comment]
There must be many more workarounds, I've merely removed the ones in clear if/else cases. E.g. there is only clearly marked in astropy/coordinates/representation.py, but there should be ton of others, too.

**Comment by pllim** [comment]
Huh...
```
    def test_count_nonzero(self):
        q1 = np.arange(9.).reshape(3, 3) * u.m
        o = np.count_nonzero(q1)
        assert type(o) is not u.Quantity
        assert o == 8
        o = np.count_nonzero(q1, axis=1)
        # Returns integer Quantity with units of m
>       assert type(o) is np.ndarray
E       AssertionError: assert <class 'astropy.units.quantity.Quantity'> is <class 'numpy.ndarray'>
E        +  where <class 'astropy.units.quantity.Quantity'> = type(<Quantity [2, 3, 3] m>)
E        +  and   <class 'numpy.ndarray'> = np.ndarray
```

Why not this?
```python
assert isinstance(o, np.ndarray)
```

**Comment by bsipocz** [comment]
> Why not this?
`assert isinstance(o, np.ndarray)`

🤷‍♀ I'll change it, but @mhvk do let me know if there was a reason for doing `type()` assertion instead. (Well, the only thing I can think of is that `isinstance()` would return True for subclasses, too, while the current check is more explicit and only accepts ndarrays)

**Comment by mhvk** [comment]
The idea is that we want to ensure that something is a real `ndarray` and not `Quantity` (which is an `ndarray` subclass. That said, why does this break...  I'll have a look later (likely not before weekend - climate strike here tomorrow).

**Comment by bsipocz** [comment]
Easily possible that I messed up one of the conditionals, will have a look. There wasn't an overall consistency package wide, sometimes it was testing for `NUMPY_LT_1_*` sometimes for `not NUMPY_LT_1_*`.

**Comment by bsipocz** [comment]
Yes, indeed, this was my mistake. Will add a commit for the fix as well as for the other failures shortly.

**Comment by bsipocz** [comment]
With the exception of the dockerized matplotlib tests, I hope everything else is fine now. @astrofrog said on slack that he would generate new containers, so ideally this should only be merged once those are ready.

File: astropy/coordinates/tests/test_frames.py [hunk 1]
```diff
@@ -187,27 +186,19 @@ def test_frame_repr():
     i3 = ICRS(ra=1*u.deg, dec=2*u.deg, distance=3*u.kpc)
 
     assert repr(i2) == ('<ICRS Coordinate: (ra, dec) in deg\n'
-                        '    ({})>').format(' 1.,  2.' if NUMPY_LT_1_14
-                                             else '1., 2.')
+                        '    ({})>').format('1., 2.')
```
Comment (mhvk) [inline astropy/coordinates/tests/test_frames.py]: Here and below, the change is of course is correct, but there isn't really a reason any more to even format the data: just insert `1., 2.` in the string proper...

File: astropy/modeling/utils.py [hunk 2]
```diff
@@ -590,7 +587,7 @@ def array_repr_oneline(array):
     """
     Represents a multi-dimensional Numpy array flattened onto a single line.
     """
-    sep = ',' if NUMPY_LT_1_14 else ', '
+    sep = ', '
     r = np.array2string(array, separator=sep, suppress_small=True)
```
Comment (mhvk) [inline astropy/modeling/utils.py]: I'd just put `separator=', '` inside the call.

File: astropy/units/quantity.py [hunk 3]
```diff
@@ -1238,7 +1206,7 @@ def __str__(self):
 
     def __repr__(self):
         prefixstr = '<' + self.__class__.__name__ + ' '
-        sep = ',' if NUMPY_LT_1_14 else ', '
+        sep = ', '
         arrstr = np.array2string(self.view(np.ndarray), separator=sep,
```
Comment (mhvk) [inline astropy/units/quantity.py]: As above, just `separator=', '`

File: astropy/units/quantity_helper/function_helpers.py [hunk 4]
```diff
@@ -483,8 +481,7 @@ def where(condition, *args):
 
 
 # Quantile was only introduced in numpy 1.15.
```
Comment (mhvk) [inline astropy/units/quantity_helper/function_helpers.py]: This comment reminds me that there are similar ones in `helpers.py` where for some reason `NUMPY_LT_*` is not used. If you are up to it, do change that too, but feel free to tell me to do it (and I'll make a PR to yours).

File: docs/table/operations.rst [hunk 5]
```diff
@@ -280,22 +280,20 @@ it is automatically ignored from aggregation.
...
-  ---- ------ -----
-  M101 13.725  15.0
-   M31   17.4  17.0
-   M82   15.5  15.7
-
-.. above tests skipped as results look different in "not NUMPY_LT_1_14".
+  >>> print(obs_by_name['mag_b'].groups.aggregate(np.mean))  # doctest: +FLOAT_CMP
+        mag_b
+  ------------------
+  15.000000000000002
```
Comment (mhvk) [inline docs/table/operations.rst]: Ha, now I see why you were reminded of #6962...

File: astropy/coordinates/tests/test_frames.py (hunk 1 above)
Comment (bsipocz) [reply astropy/coordinates/tests/test_frames.py]: good point

**Comment by astrofrog** [comment]
I'll work on the docker images tomorrow

**Comment by astrofrog** [comment]
@bsipocz - I made new versions of the docker images and pushed some commits to your branch.

**Comment by astrofrog** [comment]
The Matplotlib 2.2.2 failure is due to Matplotlib using a function deprecated in Numpy 1.16. I'm a bit surprised that the 2.1.2 build passes though

**Comment by astrofrog** [comment]
We could either explicitly ignore that deprecation warning, or we could drop support for Matplotlib<3 too...

**Comment by bsipocz** [comment]
> We could either explicitly ignore that deprecation warning, or we could drop support for Matplotlib<3 too

Actually I think that may make some sense. At least we can stop testing with versions older than 3, even if not yet making incompatible changes.

**Comment by astrofrog** [comment]
I'll try now and see if I can suppress the warning

**Comment by astrofrog** [comment]
After discussing with @eteq, for now we can just ignore that specific numpy deprecation warning when using old versions of Matplotlib.

It looks like there are still some comments from @mhvk that need addressing before this is merged, so removing the merge-when-ci-passes label

**Comment by bsipocz** [comment]
And now there are weird failures for the circleCI tests 🤷‍♀

**Comment by bsipocz** [comment]
Oh, damn. Unfortunately I force pushed on @astrofrog commits rather than first pulling them to local :( Will try to get them back now from his fork...

**Comment by bsipocz** [comment]
@astrofrog - did you have more than one commits? It doesn't seem to work, but that's being said now it doesn't seem to be mpl related, but some other weird numpy issue. I've amended my first commit with the changes to ` setup.cfg`, maybe that triggered the new failures?

**Comment by bsipocz** [comment]
somehow the mpl tests now  pick up np 1.17 rather than 1.16 which was the case at the last time they passed.

**Comment by astrofrog** [comment]
I had another commit to update the circle config, will push again

**Comment by astrofrog** [comment]
I had another commit to update the circle config, will push again

**Comment by bsipocz** [comment]
@mhvk - my pleasure, removing workarounds is a good task to do :)

**Pull request merged**